dipy/align/tests/test_vector_fields.py

```diff
@@ -1502,6 +1505,153 @@ def test_reorient_random_vector_fields():
+    expected[...,1] = 2*c*sample[:,1] + b*sample[:,0]
+    # Get the numerical gradient with the implementation under test
+    sp_to_grid = np.linalg.inv(T)
+    img_spacing = np.ones(2)
+    actual, inside = vfu.sparse_gradient(img, sp_to_grid, img_spacing, sample)
+    diff = np.abs(expected - actual).mean(1) * inside
+    # The finite differences are really not accurate, especially with float32
+    assert_equal(diff.max() < 1e-3, True)
+    # Verify exception is raised when passing invalid affine or spacings
+    invalid_affine=np.eye(2)
+    invalid_spacings = np.ones(1)
+    assert_raises(ValueError, vfu.sparse_gradient, img, invalid_affine,
+                  img_spacing, sample)
+    assert_raises(ValueError, vfu.sparse_gradient, img, sp_to_grid,
+                  invalid_spacings, sample)
+
+    # Test dense gradient
+    # Compute the analytical gradient at all points
+    expected = np.ndarray(shape + (2,), dtype=np.float32)
+    expected[...,0] = 2*a*TX[...,0] + b*TX[...,1]
+    expected[...,1] = 2*c*TX[...,1] + b*TX[...,0]
+    # Get the numerical gradient with the implementation under test
+    sp_to_grid = np.linalg.inv(T)
+    img_spacing = np.ones(2)
+    actual, inside = vfu.gradient(img, sp_to_grid, img_spacing, shape, T)
+    diff = np.abs(expected - actual).mean(2) * inside
+    # In the dense case, we are evaluating at the exact points (sample points
+    # are not slightly moved like in the sparse case), so we have more precision
+    assert_equal(diff.max() < 1e-5, True)
+    # Verify exception is raised when passing invalid affine or spacings
+    assert_raises(ValueError, vfu.gradient, img, invalid_affine, img_spacing,
+                  shape, T)
+    assert_raises(ValueError, vfu.gradient, img, sp_to_grid, img_spacing,
+                  shape, invalid_affine)
+    assert_raises(ValueError, vfu.gradient, img, sp_to_grid, invalid_spacings,
+                  shape, T)
+
+
+def test_gradient_3d():
+    np.random.seed(3921116)
+    shape = (25, 32, 15)
+    # Create grid coordinates
+    x_0 = np.asarray(range(shape[0]))
+    x_1 = np.asarray(range(shape[1]))
+    x_2 = np.asarray(range(shape[2]))
+    X = np.ndarray(shape+(4,), dtype = np.float64)
+    O = np.ones(shape)
+    X[..., 0]= x_0[:, None, None] * O
+    X[..., 1]= x_1[None, :, None] * O
+    X[..., 2]= x_2[None, None, :] * O
+    X[..., 3]= 1
+
+    transform = regtransforms[('RIGID', 3)]
+    theta = np.array([0.1, 0.05, 0.12, -12.0, -15.5, -7.2])
+    T = transform.param_to_matrix(theta)
+
+    TX = X.dot(T.T)
+    # Eval an arbitrary (known) function at TX
+    # f(x, y, z) = ax^2 + by^2 + cz^2 + dxy + exz + fyz
+    # df/dx = 2ax + dy + ez
+    # df/dy = 2by + dx + fz
+    # df/dz = 2cz + ex + fy
+    a, b, c = 2e-3, 3e-3, 1e-3
+    d, e, f = 1e-3, 2e-3, 3e-3
+    img = a*TX[...,0]**2 + b*TX[...,1]**2 + c*TX[...,2]**2 + \
+          d*TX[...,0]*TX[...,1] + e*TX[...,0]*TX[...,2] + f*TX[...,1]*TX[...,2]
```

Comment:
pep8

I would consider using an editor like spyder etc. who can tell you when you have pep8 issues.